base_location/tests/test_base_location.py

```diff
@@ -0,0 +1,315 @@
+        })
+        company._inverse_city_id()
+        company._inverse_zip_id()
+        self.assertEqual(company.zip_id, company.partner_id.zip_id)
+        self.assertEqual(company.city_id, company.partner_id.city_id)
+
+    def test_onchange_company_city_id_completion(self):
+        """Test city auto-completion when changing zip in a company"""
+        company = self.env['res.company'].new({'name': 'Test'})
+        city_zip2 = self.env['res.city.zip'].create(
+            self.values_city_zip2())
+        company.zip_id = city_zip2
+        company._onchange_zip_id()
+        self.assertEqual(company.city_id, city_zip2.city_id)
+
+    def test_constrains_partner_01(self):
+        """Test partner 1 constraints"""
+        city_zip2 = self.env['res.city.zip'].create(
+            self.values_city_zip2())
+        with self.assertRaises(ValidationError):
+            self.env['res.partner'].create({
+                'name': 'P1',
+                'zip_id': city_zip2.id,
+            })
+
+    def test_constrains_partner_country(self):
+        """Test partner country constraints"""
+        city_zip2 = self.env['res.city.zip'].create(
+            self.values_city_zip2())
+        partner = self.env['res.partner'].create({
+            'name': 'P1',
+            'zip_id': city_zip2.id,
+            'country_id': city_zip2.city_id.country_id.id,
+            'state_id': city_zip2.city_id.state_id.id,
+            'city_id': city_zip2.city_id.id,
+        })
+
+        with self.assertRaises(ValidationError):
+            partner.country_id = self.ref('base.ch')
+
+    def test_constrains_partner_state(self):
+        """Test partner state constraints"""
+        city_zip2 = self.env['res.city.zip'].create(
+            self.values_city_zip2())
+        partner = self.env['res.partner'].create({
+            'name': 'P1',
+            'zip_id': city_zip2.id,
+            'country_id': city_zip2.city_id.country_id.id,
+            'state_id': city_zip2.city_id.state_id.id,
+            'city_id': city_zip2.city_id.id,
+        })
+
+        with self.assertRaises(ValidationError):
+            partner.state_id = self.state_vd.id
+
+    def test_constrains_partner_city(self):
+        """Test partner city constraints"""
+        city_zip2 = self.env['res.city.zip'].create(
+            self.values_city_zip2())
+        partner = self.env['res.partner'].create({
+            'name': 'P1',
+            'zip_id': city_zip2.id,
+            'country_id': city_zip2.city_id.country_id.id,
+            'state_id': city_zip2.city_id.state_id.id,
+            'city_id': city_zip2.city_id.id,
+        })
+
+        with self.assertRaises(ValidationError):
+            partner.city_id = self.city_lausanne
+
+    def test_partner_onchange_country(self):
+        """Test partner onchange country_id"""
+        country_es = self.browse_ref('base.es')
+        country_es.enforce_cities = True
+        city_zip1 = self.env['res.city.zip'].create(
```

Comment:
seems like you need these city_zip1/2 in every tests: create them in setUpClass.
Is useless and resource / time consuming when running tests ;)